File: game/levels/components.py
import json


def enable(ctx, component):
    c = ctx.get_target(component)
    if c["state"] != "active":
        c["state"] = "active"
        ctx.exec(c, "enable")


def disable(ctx, component):
    c = ctx.get_target(component)
    if c["state"] != "inactive":
        c["state"] = "inactive"
        ctx.exec(c, "disable")


def duplicate(ctx, component, duplicate_id):
    c = json.loads(json.dumps(ctx.get_target(component)))
    c["id"] = duplicate_id
    ctx.components.append(c)
    return c


def change(ctx, target, path, value):
    subpaths = path.split(".")

    if subpaths[0] == "state":
        raise Exception("Don't change state with change(), use enable() or disable()")

    t = ctx.get_target(target)
    for subpath in subpaths[:-1]:
        t = t[subpath]
    t[subpaths[-1]] = value

File: lib/levels-functions/components.js
export const enable = (component) => {
  const c = this.get_target(component);
  if (c.state !== 'active') {
    c.state = 'active';
    this.mark_dirty(c);
    this.exec(c, 'enable');
  }
};

export const disable = (component) => {
  const c = this.get_target(component);
  if (c.state !== 'inactive') {
    c.state = 'inactive';
    this.mark_dirty(c);
    this.exec(c, 'disable');
  }
};

export const duplicate = (component, duplicate_id) => {
  const c = JSON.parse(JSON.stringify(this.get_target(component)));
  c.id = duplicate_id;
  this.mark_dirty(c);
  this.components.push(c);
  return c;
};

export const change = (target, path, value) => {
  const subpaths = path.split('.');

  if (subpaths[0] == 'state') {
    throw new Error("Don't change state with change(), use enable() or disable()");
  }
  const c = this.get_target(target);
  this.mark_dirty(c);
  let t = c;

  const lastItem = subpaths.pop();
  for (let subpath of subpaths) {
    t = t[subpath];
  }
  t[lastItem] = value;

  if (subpaths[0] == 'visibility') {
    this.mark_dirty(c); // needs to be marked dirty again, in case we changed visibility for some players
  }
};
